Implement heartbeat metrics

roidelapluie opened the pull request:
pt-heartbeat measures actual replication lab on a mysql server. It does
not rely on data stored in SHOW SLAVE STATUS, as it can be unreliable.

pt-heartbeat website:
https://www.percona.com/doc/percona-toolkit/2.2/pt-heartbeat.html

Signed-off-by: Julien Pivotto <roidelapluie@inuits.eu>

brian-brazil commented:
How many different heartbeat implementations are there out there?

roidelapluie commented:
pt-heartbeat has been there for a very long time; it is widely used.

grypyrg commented:
@brian-brazil : `pt-heartbeat` is the most widely used (if not only) alternative tool to measure replication lag properly.
(Of course, there are many companies that do heartbeats using a custom built solution)

Given these 2 new metrics, can we easily get the time difference with them?
Maybe we should also calculate in `mysqld_exporter` itself already how much the difference is and write it to a different metric to make it easier to report the replication lag.

SuperQ commented:
Could be interesting.

In what way is `SHOW SLAVE STATUS` unreliable?

Monitoring for slave lag with Prometheus

SuperQ commented:
Argh, wrong button. And no way to edit on mobile.

grypyrg commented:
`SHOW SLAVE STATUS` is unreliable when:
- parallel replication is used (`slave_parallel_workers>0`). several metrics are only updated in regular intervals (https://dev.mysql.com/doc/refman/5.7/en/replication-options-slave.html#option_mysqld_slave-checkpoint-group & https://dev.mysql.com/doc/refman/5.7/en/replication-options-slave.html#option_mysqld_slave-checkpoint-period
- when IO thread is not catching up with the master (due to limited bandwith), and the sql thread has caught up with the IO thread, `seconds_behind_master` will reported as `0`
- in multi-tiered architectures, replication lag is always the timestamp from where the query was originally executed. with `pt-heartbeat` you can run the heartbeat on the intermediate server and get replication lag for that tier only
- when timezones are different between servers you can have for example `+3600` or `-3600` report seconds replication lag.
- https://bugs.mysql.com/bug.php?id=80084
- ...

roidelapluie commented:
Fixed:
- added _seconds suffix as recommended in Prometheus documentation.
- added the actual lag time
- rounded to integers. Makes no sense to go below the second as pt-heartbeat only updates the table once per second.

SuperQ commented:
@grypyrg Awesome, thanks for documenting those issues here.

SuperQ commented:
Even with second-only precision hearbeats, it's still worth keeping the float data as is valid.  Especially if `pt-heartbeat` attempts to write on-the-second and not just `sleep(1)`

Maybe for compatibility, we should implement a `pt-hearbeat` compatible writer and include it in the `mysqld_exporter` repo.  We could also implement it to write out millisecond precision timestamps.

It's really too bad that `pt-heartbeat` uses a varchar for the timestamp instead of a `datetime(6)`.

SuperQ commented on collector/pt_heartbeat.go:
```diff
@@ -0,0 +1,95 @@
+	ptHeartbeat = "pt_heartbeat"
+	// pt-heartbeat query
+	// the second column allows gets the server timestamp at the exact same time the query is run
+	ptHeartbeatQuery = "SELECT UNIX_TIMESTAMP(ts), UNIX_TIMESTAMP(CONCAT(DATE(NOW()), ' ', CURTIME(6))), server_id from %s"
```
This can be simplified with `UNIX_TIMESTAMP(NOW(6))`.  I tested on 5.5 and 5.6.

roidelapluie replied:
fixed

roidelapluie commented:
@SuperQ pt-heartbeat'i very initial version dates from https://sourceforge.net/p/maatkit/code/857/ 2007-09-05. The latest version is the result of almost 10 years of maturing. Older releases were using timestamps, and that got changed to VARCHAR later. I could not track back the reason.

I do agree that a Go version of that tool might be handy, but is out of scope of this pull request.

SuperQ commented on collector/pt_heartbeat.go:
```diff
@@ -0,0 +1,95 @@
+		"Timestamp of the current server.",
+		[]string{"server_id"}, nil,
+	)
+	PtHeartbeatDelayDesc = prometheus.NewDesc(
```
This metric isn't necessary (and against Prometheus design) since you can do this in PromQL.

roidelapluie replied:
@grypyrg That delay can be made at the prometheus level using https://prometheus.io/docs/querying/rules/ . Once this patch gets in, maybe we can document an example recording rile for this on a blogpost.

roidelapluie replied:
removed

SuperQ commented on README.md:
```diff
@@ -66,6 +66,8 @@ collect.perf_schema.indexiowaits                       | 5.6           | Collect
 collect.perf_schema.tableiowaits                       | 5.6           | Collect metrics from performance_schema.table_io_waits_summary_by_table.
 collect.perf_schema.tablelocks                         | 5.6           | Collect metrics from performance_schema.table_lock_waits_summary_by_table.
 collect.slave_status                                   | 5.1           | Collect from SHOW SLAVE STATUS (Enabled by default)
+collect.pt-heartbeat                                   | 5.1           | Collect from pt-heartbeat
```
I would prefer to call this `heartbeat`, as there could be other implementations.

roidelapluie replied:
renamed everywhere

SuperQ commented:
Overall, I like this idea, some comments in-line.

roidelapluie commented:
Should I rename ptHeartbeat to heartbeat everywhere or just in the CLI flags?

roidelapluie commented:
Updated after latest comments.

@SuperQ do you want more documintatior for this in README?

SuperQ commented:
@roidelapluie Totally agree, re-implementing is out of scope for this feature.  I was mostly just thinking out loud.

Yes, please remove `pt` from everything.  But add a note and a link to `pt-heartbeat` in the README that "this is a supported implementation of heartbeat checking".

roidelapluie commented on collector/heartbeat.go:
```diff
@@ -0,0 +1,88 @@
+	// timestamps. %s will be replaced by the table name.
+	// The second column allows gets the server timestamp at the exact same
+	// time the query is run.
+	heartbeatQuery = "SELECT UNIX_TIMESTAMP(ts), UNIX_TIMESTAMP(NOW(6)), server_id from `%s`"
```
@SuperQ how can we prevent this to turn into sql injection?

roidelapluie replied:
Apparently not much https://github.com/golang/go/issues/18478

roidelapluie replied:
split in two: database and table, for better escaping

grypyrg commented:
@roidelapluie : `pt-heartbeat` does support subsecond checking. See https://www.percona.com/doc/percona-toolkit/2.1/pt-heartbeat.html#cmdoption-pt-heartbeat--interval 
```
The minimum (fastest) interval is 0.01, and the maximum precision is two decimal places, 
so 0.015 will be rounded to 0.02.
```

grypyrg commented:
And maybe fetch another column to precalculate the `seconds_lag`

```
SELECT UNIX_TIMESTAMP(ts) AS executed_time,
       UNIX_TIMESTAMP(NOW(6)) AS server_time,
       UNIX_TIMESTAMP(NOW(6)) - UNIX_TIMESTAMP(ts) AS seconds_lag,
       server_id
FROM `%s`
```

Example output (to also show it's subsecond):
```
+-------------------+-------------------+-------------+-----------+
| executed_time     | server_time       | seconds_lag | server_id |
+-------------------+-------------------+-------------+-----------+
| 1487668183.001080 | 1487668278.442121 |   95.441041 |         1 |
+-------------------+-------------------+-------------+-----------+
```

roidelapluie commented:
@grypyrg According to @grypyrg the lag should be computed at the promsql level.

roidelapluie commented:
Updated with all the commands taken in consideration; also split database and table parameters so we call better escape them.

SuperQ commented:
Maybe make a parameter for the column?

roidelapluie commented:
Two parameters then. Is that complexity needed for an initial release?

SuperQ commented:
No, not necssary, it was just an idea.  I think it's fine to keep the columns hard-coded.

SuperQ commented on collector/heartbeat.go:
```diff
@@ -0,0 +1,88 @@
+	}
+	defer heartbeatRows.Close()
+
+	var now sql.RawBytes
```
This should be:
```
var (
  now, ts  sql.RawBytes
  serverId int
)
```

roidelapluie commented:
Go nit fixed.

roidelapluie commented:
-2 now until I check my english wording

roidelapluie commented:
Wording fixed in README.md and mysqld_exporter.go, so for me it is good to merge.

SuperQ commented:
Yup, looks fine, thanks.

grypyrg commented:
Thanks @roidelapluie !

roidelapluie commented:
Fer reference, promsql query:

mysql_heartbeat_now_timestamp_seconds - on (server_id) mysql_heartbeat_stored_timestamp_seconds

SuperQ commented:
Normally Prometheus will be able to match up the labels and you don't need to specify the `on (server_id)`.

roidelapluie commented:
Oh thanks. Nice trick.
